src/lib.rs

```diff
@@ -14,38 +15,77 @@ fn derive_abomonation(mut s: synstructure::Structure) -> proc_macro2::TokenStrea
     });
 
     let entomb = s.each(|bi| quote! {
-        ::abomonation::Abomonation::entomb(#bi, _write)?;
+        ::abomonation::Entomb::entomb(#bi, _write)?;
     });
 
-    let extent = s.each(|bi| quote! {
-        sum += ::abomonation::Abomonation::extent(#bi);
-    });
+    // T::alignment() is the max of mem::align_of<T> and the U::alignment()s of
+    // every U type used as a struct member or inside of an enum variant (which
+    // includes the alignment of recursively abomonated data)
+    //
+    // Unfortunately, we cannot use synstructure's nice `fold()` convenience
+    // here because it's based on generating match arms for a `self` value,
+    // whereas here we're trying to implement an inherent type method without
+    // having such a `self` handy.
+    //
+    // We can, however, use Structure::variants() and VariantInfo::bindings()
+    // to enumerate all the types which _would appear_ in such match arms'
+    // inner bindings.
+    //
+    let mut alignment = vec![
+        quote!(let mut align = ::std::mem::align_of::<Self>();)
+    ];
+    let mut probed_types = HashSet::new();
+    for variant_info in s.variants() {
+        for binding_info in variant_info.bindings() {
+            let binding_type = &binding_info.ast().ty;
+            // Do not query a type's alignment() multiple times
+            if probed_types.insert(binding_type) {
+                alignment.push(
+                    quote!(align = align.max(<#binding_type as ::abomonation::Entomb>::alignment());)
+                );
+            }
+        }
+    }
+    alignment.push(quote!(align));
```

Comment:
Abomonation API change from https://github.com/TimelyDataflow/abomonation/pull/33: after this PR, abomonation will provide full support for correctly aligning data in memory... which requires from support from the Abomonation trait implementors.